compiler/bindings/python/iree/build/executor.py

```diff
@@ -0,0 +1,467 @@
+        stack.add(initial_dep)
+        for producer_dep in initial_dep.deps:
+            self._add_dep(producer_dep, initial_dep, stack)
+
+    def _add_dep(
+        self,
+        producer_dep: BuildDependency,
+        consumer_dep: BuildDependency,
+        stack: set[BuildDependency],
+    ):
+        if producer_dep in stack:
+            raise RuntimeError(
+                f"Circular dependency: '{producer_dep}' depends on itself: {stack}"
+            )
+        plist = self.producer_graph.get(producer_dep)
+        if plist is None:
+            plist = []
+            self.producer_graph[producer_dep] = plist
+        plist.append(consumer_dep)
+        next_stack = set(stack)
+        next_stack.add(producer_dep)
+        if producer_dep.deps:
+            # Intermediate dep.
+            for next_dep in producer_dep.deps:
+                self._add_dep(next_dep, producer_dep, next_stack)
+
+    def build(self):
+        # Build all deps until the graph is satisfied.
+        # Schedule any deps that have no dependencies to start things off.
+        for eligible_dep in self.producer_graph.keys():
+            if len(eligible_dep.deps) == 0:
+                self._schedule_action(eligible_dep)
+                self.in_flight_deps.add(eligible_dep)
+
+        while self.producer_graph:
+            print(f"Servicing {len(self.producer_graph)} outstanding tasks")
+            self._service_graph()
+
+    def _service_graph(self):
+        completed_deps: set[BuildDependency] = set()
+        try:
+            for completed_fut in concurrent.futures.as_completed(
+                (d.future for d in self.in_flight_deps), 0
+            ):
+                completed_dep = completed_fut.result()
+                assert isinstance(completed_dep, BuildDependency)
+                print(f"Completed {completed_dep}")
+                completed_deps.add(completed_dep)
+        except TimeoutError:
+            pass
+
+        # Purge done from in-flight list.
+        self.in_flight_deps.difference_update(completed_deps)
+
+        # Schedule any available.
+        for completed_dep in completed_deps:
+            ready_list = self.producer_graph.get(completed_dep)
+            if ready_list is None:
+                continue
+            del self.producer_graph[completed_dep]
+            for ready_dep in ready_list:
+                self._schedule_action(ready_dep)
+                self.in_flight_deps.add(ready_dep)
+
+        # Do a blocking wait for at least one ready.
+        concurrent.futures.wait(
+            (d.future for d in self.in_flight_deps),
+            return_when=concurrent.futures.FIRST_COMPLETED,
+        )
+
+    def _schedule_action(self, dep: BuildDependency):
+        if dep.is_scheduled:
+            return
+        if isinstance(dep, BuildAction):
+
+            def invoke():
+                dep.invoke()
+                return dep
+
+            print(f"Scheduling action: {dep}")
```

Comment:
May want to use `logging` here instead of `print`. Also, since this is a full build dependency graph executor, I might find it useful to debug the build graph if it grows complex for some tasks. Could consider more logging or `.dot` file generation

Reply:
It's pretty traditional for build tools to emit unadorned output. Agreed that it needs some IO helper but logging probably isn't the thing.